### DISCUSSION TITLE
can we cache Docker layer of Super-linter image in Github action

### DISCUSSION THREAD

**Discussion by ABRAJ11** [super-linter/super-linter#3101] [General]
Thanks for Awesome  action . Currently the Docker pull takes around  1-2 min in pulling docker image . can we enable caching of super-linter container image to enable fast execution of workflow

**Comment by osfanbuff63** [comment]
I don't think that you can right now - but this is a much-needed feature since the docker pull takes 2-3x more (at least) vs the rest of the workflow.

**Reply by Shahard2** [reply]
Any updates on this one?

**Reply by BretFisher** [reply]
If we're talking about public runners, I think what you're asking is "can the public runner VMs cache the super-linter images in docker's image cache when they are built?"

In that case, you'd need to request it being added to the image list for Linux VM builds, which currently [only caches distro-specific images](https://github.com/actions/runner-images/blob/main/images/linux/toolsets/toolset-2204.json#L210-L231) AFAIK.

**Reply by YElyousfi** [reply]
What about using `actions/cache` just to speed things up on a workflow level?

**Reply by BretFisher** [reply]
Hmm, I doubt that could work in the current model, and also unlikely it would improve speed.

Speed: You're either downloading the image from a registry during the docker step, or downloading the cache during a different step. Also, container registries are optimized for multiple streams (one per layer) so they download faster in tandem. I doubt the cache would do the same for a stored file system.

Compatibility: You would need to reconfigure the Docker Engine (installed already on the public runners) to see a different location (the cache dir) in order to store images there. Docker can't store image layers in a different place at will. This question about actions/cache+docker has been [asked and answered multiple times](https://github.com/actions/cache/issues?q=in:title+docker), and the answer is that cache doesn't support docker image cache.

Workaround: There is a solution, though, and it's to setup your own private runners and cache all the images you'd need on those runners during initial setup. I've seen setups that [autoscale their own runners](https://docs.github.com/en/actions/hosting-your-own-runners/autoscaling-with-self-hosted-runners) (so none are running when you aren't using them) and they seed an image list during setup.

**Comment by lkraav** [comment]
Is more caching opportunity available with self-hosted runners?

**Reply by BretFisher** [reply]
Yes, if your self-hosted runner is long-lived, and you don't prune docker's image cache, then the super-linter image would be cached.

Note that most of the self-hosted runner cloud solutions destroy idle runners, causing similar issues to public runners where the docker image cache is blank on each Job.

Also, for long-running self-hosted runners, you'll quickly find that image caching will fill the disk, and will require something like a daily `docker image prune -af --filter "until=24h"` to remove old images. This would also cause the removal of super-linter, but at least it would only be pulled once a day 😉.

**Comment by ferrarimarco** [comment]
I think this is more on the GitHub Actions runners side, rather than on super-linter side. I would open an issue in the hosted runners images repo.
